src/category_theory/with_term.lean

```diff
@@ -0,0 +1,322 @@
+| star (of X) _ := λ f g, pempty.elim f
+| _ star (of Y) := λ f g, pempty.elim g
+| star star star := λ _ _, punit.star
+
+instance : category.{v} (with_term C) :=
+{ hom := λ X Y, hom X Y,
+  id := λ X, id _,
+  comp := λ X Y Z f g, comp f g }
+
+/-- The inclusion from `C` into `with_term C`. -/
+def incl : C ⥤ (with_term C) :=
+{ obj := of,
+  map := λ X Y f, f }
+
+instance : full (incl : C ⥤ _) :=
+{ preimage := λ X Y f, f }
+
+instance : faithful (incl : C ⥤ _) := {}
+
+/-- Map `with_term` with respect to a functor `F : C ⥤ D`. -/
+def map {D : Type*} [category D] (F : C ⥤ D) : with_term C ⥤ with_term D :=
+{ obj := λ X,
+    match X with
+    | of x := of $ F.obj x
+    | star := star
+    end,
+  map := λ X Y f,
+    match X, Y, f with
+    | of x, of y, f := F.map f
+    | of x, star, punit.star := punit.star
+    | star, star, punit.star := punit.star
+    end }
+
+instance {X : with_term C} : unique (X ⟶ star) :=
+{ default :=
+    match X with
+    | of x := punit.star
+    | star := punit.star
+    end,
+  uniq := by tidy }
+
+/-- `with_term.star` is terminal. -/
+def star_terminal : limits.is_terminal (star : with_term C) :=
+limits.is_terminal.of_unique _
+
+/-- Lift a functor `F : C ⥤ D` to `with_term C ⥤ D`, given a terminal object of `D`. -/
+@[simps]
+def lift {D : Type*} [category D] {Z : D} (hZ : limits.is_terminal Z) (F : C ⥤ D) :
+  (with_term C) ⥤ D :=
+{ obj := λ X,
+    match X with
+    | of x := F.obj x
+    | star := Z
+    end,
+  map := λ X Y f,
+    match X, Y, f with
+    | of x, of y, f := F.map f
+    | of x, star, punit.star := hZ.from _
+    | star, star, punit.star := hZ.from _
+    end,
+  map_id' := begin
+    rintros (X|X),
+    apply F.map_id,
+    apply hZ.hom_ext,
+  end,
+  map_comp' := begin
+    rintros (X|X) (Y|Y) (Z|Z) f g,
+    any_goals {apply hZ.hom_ext},
+    tidy,
+  end }
+
+/-- The isomorphism between `incl ⋙ lift _ F` with `F`. -/
+@[simps]
+def incl_lift {D : Type*} [category D] {Z : D} (hZ : limits.is_terminal Z) (F : C ⥤ D) :
+  incl ⋙ lift hZ F ≅ F :=
+{ hom := { app := λ X, 𝟙 _ },
+  inv := { app := λ X, 𝟙 _ } }
+
+/-- The uniqueness of `lift _ F`. -/
+@[simps]
+def lift_unique {D : Type*} [category D] {Z : D} (hZ : limits.is_terminal Z) (F : C ⥤ D)
+  (G : with_term C ⥤ D) (h : incl ⋙ G ≅ F) (hG : G.obj star ≅ Z) : G ≅ lift hZ F :=
+{ hom :=
+  { app := λ X,
+      match X with
+      | of x := h.hom.app x
+      | star := hG.hom
+      end,
+    naturality' := begin
+      rintros (X|X) (Y|Y) f,
+      any_goals {apply hZ.hom_ext},
+      apply h.hom.naturality,
+      cases f,
+    end },
+  inv :=
+  { app := λ X,
+      match X with
+      | of x := h.symm.hom.app x
+      | star := hG.symm.hom
+      end,
+    naturality' := begin
+      rintros (X|X) (Y|Y) f,
+      any_goals {apply (limits.is_terminal.of_iso hZ hG.symm).hom_ext},
+      apply h.symm.hom.naturality,
+      cases f,
+    end },
+  hom_inv_id' := by {ext (X|X), tidy},
+  inv_hom_id' := by {ext (X|X), tidy} }
```

Comment:
This can be golfed to
```suggestion
nat_iso.of_components (λ X, 
  match X with
  | of x := h.app x
  | star := hG
  end)
  (begin
    rintros (X|X) (Y|Y) f,
    any_goals { apply hZ.hom_ext },
    apply h.hom.naturality,
    cases f,
  end)
```